subcerts.go

```diff
@@ -0,0 +1,508 @@
+	// The signature scheme used to sign the credential.
+	Scheme SignatureScheme
+
+	// The credential's delegation.
+	Signature []byte
+}
+
+// NewDelegatedCredential creates a new delegated credential using `cert` for
+// delegation. It generates a public/private key pair for the provided signature
+// algorithm (`scheme`), validity interval (defined by `cert.Leaf.notBefore` and
+// `validTime`), and TLS version (`vers`), and signs it using `cert.PrivateKey`.
+func NewDelegatedCredential(cert *Certificate, scheme SignatureScheme, validTime time.Duration, vers uint16) (*DelegatedCredential, error) {
+	// The granularity of DC validity is seconds.
+	validTime = validTime.Round(time.Second)
+
+	// Parse the leaf certificate if needed.
+	var err error
+	if cert.Leaf == nil {
+		cert.Leaf, err = x509.ParseCertificate(cert.Certificate[0])
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	// Check that the leaf certificate can be used for delegation.
+	if !canDelegate(cert.Leaf) {
+		return nil, errNoDelegationUsage
+	}
+
+	// Extract the delegator signature scheme rom the certificate.
+	var delegatorScheme SignatureScheme
+	switch sk := cert.PrivateKey.(type) {
+	case *ecdsa.PrivateKey:
+
+		// Set scheme
+		pk := sk.Public().(*ecdsa.PublicKey)
+		curveName := pk.Curve.Params().Name
+		certAlg := cert.Leaf.SignatureAlgorithm
+		if certAlg == x509.ECDSAWithSHA256 && curveName == "P-256" {
+			delegatorScheme = ECDSAWithP256AndSHA256
+		} else if certAlg == x509.ECDSAWithSHA384 && curveName == "P-384" {
+			delegatorScheme = ECDSAWithP384AndSHA384
+		} else if certAlg == x509.ECDSAWithSHA512 && curveName == "P-521" {
+			delegatorScheme = ECDSAWithP521AndSHA512
+		} else {
+			return nil, fmt.Errorf(
+				"using curve %s for %s is not supported",
+				curveName, cert.Leaf.SignatureAlgorithm)
+		}
+
+	default:
+		return nil, fmt.Errorf("unsupported delgation key type: %T", sk)
+	}
+
+	// Generate a new key pair.
+	var sk crypto.PrivateKey
+	var pk crypto.PublicKey
+	switch scheme {
+	case ECDSAWithP256AndSHA256,
+		ECDSAWithP384AndSHA384,
+		ECDSAWithP521AndSHA512:
+
+		sk, err = ecdsa.GenerateKey(getCurve(scheme), rand.Reader)
+		if err != nil {
+			return nil, err
+		}
+		pk = sk.(*ecdsa.PrivateKey).Public()
+
+	default:
+		return nil, fmt.Errorf("unsupported signature scheme: 0x%04x", scheme)
+	}
+
+	// Prepare the credential for digital signing.
+	hash := getHash(delegatorScheme)
+	cred := &credential{validTime, pk, scheme}
+	in, err := prepareDelegation(hash, cred, cert.Leaf.Raw, delegatorScheme, vers)
+	if err != nil {
+		return nil, err
+	}
+
+	// Sign the creential.
+	var sig []byte
+	switch sk := cert.PrivateKey.(type) {
+	case *ecdsa.PrivateKey:
+		opts := crypto.SignerOpts(hash)
+		sig, err = sk.Sign(rand.Reader, in, opts)
+		if err != nil {
+			return nil, err
+		}
+	default:
+		return nil, fmt.Errorf("unsupported delgation key type: %T", sk)
+	}
+
+	return &DelegatedCredential{
+		nil, validTime, pk, sk, scheme, delegatorScheme, sig}, nil
```

Comment:
Long-form would be good here. It handles the wrap problem, and makes what-is-what clearer to me,
```
&DelegatedCredential{
    ValidTime: validTime,
 ...
```

Reply:
Also, a lot of this file is big if-tables and switch statements and feels repetitive. I would think about how to reduce this, or compartmentalize it. Not necessary to do, just a thought.

Reply:
^Ack.

Reply:
^^Agreed. Done.